src/lib/transition_router/initial_validator.ml

```diff
@@ -1,80 +1,93 @@
 open Async_kernel
-open Protocols.Coda_pow
 open Pipe_lib.Strict_pipe
 open Coda_base
 open Coda_state
 
-module type Inputs_intf = sig
-  include Transition_frontier.Inputs_intf
-
-  module State_proof :
-    Proof_intf with type input := Protocol_state.Value.t and type t := Proof.t
+module Make (Inputs : Transition_frontier.Inputs_intf) = struct
+  open Inputs
 
-  module Time : Time_intf
+  type validation_error =
+    [ `Invalid_time_received of [`Too_early | `Too_late of int64]
+    | `Invalid_proof
+    | `Verifier_error of Error.t ]
 
-  module Transition_frontier :
-    Transition_frontier_intf
-    with type state_hash := State_hash.t
-     and type external_transition_verified := External_transition.Verified.t
-     and type ledger_database := Ledger.Db.t
-     and type staged_ledger := Staged_ledger.t
-     and type transaction_snark_scan_state := Staged_ledger.Scan_state.t
-     and type staged_ledger_diff := Staged_ledger_diff.t
-     and type masked_ledger := Coda_base.Ledger.t
-     and type consensus_local_state := Consensus.Data.Local_state.t
-     and type user_command := User_command.t
-     and type diff_mutant :=
-                ( External_transition.Stable.Latest.t
-                , State_hash.Stable.Latest.t )
-                With_hash.t
-                Diff_mutant.E.t
+  type ('time_received_valid, 'proof_valid) validation_result =
+    ( ( [`Time_received] * 'time_received_valid
+      , [`Proof] * 'proof_valid
+      , [`Frontier_dependencies] * Truth.false_t
+      , [`Staged_ledger_diff] * Truth.false_t )
+      External_transition.Validation.with_transition
+    , validation_error )
+    Deferred.Result.t
 
-  module Protocol_state_validator :
-    Protocol_state_validator_intf
-    with type time := Time.t
-     and type state_hash := State_hash.t
-     and type trust_system := Trust_system.t
-     and type envelope_sender := Envelope.Sender.t
-     and type external_transition := External_transition.t
-     and type external_transition_proof_verified :=
-                External_transition.Proof_verified.t
-     and type external_transition_verified := External_transition.Verified.t
-end
-
-module Make (Inputs : Inputs_intf) :
-  Protocols.Coda_transition_frontier.Initial_validator_intf
-  with type time := Inputs.Time.t
-   and type state_hash := State_hash.t
-   and type trust_system := Trust_system.t
-   and type external_transition := Inputs.External_transition.t
-   and type external_transition_verified :=
-              Inputs.External_transition.Verified.t = struct
-  open Inputs
+  let handle_validation_error ~logger ~trust_system ~sender ~state_hash
+      (error : validation_error) =
+    match error with
+    | `Verifier_error err ->
+        Logger.error logger ~module_:__MODULE__ ~location:__LOC__
+          ~metadata:[("state_hash", State_hash.to_yojson state_hash)]
+          "Error while verifying blockchain proof for $state_hash: %s"
+          (Error.to_string_hum err) ;
+        return ()
```

Comment:
This should proooobably be a `Trust_system.Actions.Sent_invalid_proof`? Crashing the verifier isn't that different than an actual invalid proof from this node's perspective...

Reply:
fair enough